src/Core/src/Credentials/SsoCacheFileLoader.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Core\Credentials;
+
+use AsyncAws\Core\EnvVar;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+
+/**
+ * Load and parse AWS SSO cache file.
+ */
+final class SsoCacheFileLoader
+{
+    public const KEY_ACCESS_TOKEN = 'accessToken';
+    public const KEY_EXPIRES_AT = 'expiresAt';
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    public function __construct(?LoggerInterface $logger = null)
+    {
+        $this->logger = $logger ?? new NullLogger();
+    }
+
+    /**
+     * @return array<string, string>
+     */
+    public function loadSsoCacheFile(string $ssoStartUrl): array
+    {
+        $filepath = sprintf('%s/.aws/sso/cache/%s.json', $this->getHomeDir(), sha1($ssoStartUrl));
+
+        if (!@is_readable($filepath)) {
+            $this->logger->warning('The sso cache file {path} is not readable.', ['path' => $filepath]);
```

Comment:
I don't think this is the right place for this check, and opens up a race condition between here and line 42. `file_get_contents` could still fail, but we don't properly handle that.

Reply:
Example robust ways to open the file, then read it's contents can be found in the Guzzle Utils class (`tryFopen`, `tryGetContents`): https://github.com/guzzle/psr7/blob/2.5.0/src/Utils.php#L340-L434.

Reply:
What do you think about this? Would it be enough?
```php
if (!@is_readable($filepath) || false === ($content = file_get_contents($filepath))) {
```

Reply:
`file_get_contents` can also raise errors. The linked code handles them and surfaces they back to the user as an exception.

Reply:
Checking is_readable is largely not useful, because the file's readability could change in between that call and then the file_get_contents call.

Reply:
FYI the `IniFileLoader` has the same race condition:
- https://github.com/async-aws/aws/blob/master/src/Core/src/Credentials/IniFileLoader.php#L54
- https://github.com/async-aws/aws/blob/master/src/Core/src/Credentials/IniFileLoader.php#L95

Reply:
Great - this means we are justified to add a little internal class in the Core namespace to get the contents of a file or throw an exception.

Reply:
@GrahamCampbell Since the [referenced code](https://github.com/guzzle/psr7/blob/2.5.0/src/Utils.php#L340-L434) is written by yourself, would you consider to contribute it here?

Reply:
Since I didn't get an answer, I will go with following:
```php
if (false === ($contents = @file_get_contents($filepath))) {
```
I don't fell well to copy the referenced code when the original author is involved here 😉

Reply:
are you OK with this change @GrahamCampbell

Reply:
Yeh, we could also just "fix it in post", since this is an existing bug, and nobody seems to mind it so far. Are you OK if we make an issue for it, assign me to it, then I'll get to it when I have some time.